src/gfn/samplers.py

```diff
@@ -264,3 +268,275 @@ def sample_trajectories(
         )
 
         return trajectories
+
+
+class LocalSearchSampler(Sampler):
+    """Sampler equipped with local search capabilities.
+    The local search operation is based on back-and-forth heuristic, first proposed
+    by Zhang et al. 2022 (https://arxiv.org/abs/2202.01361) for negative sampling
+    and further explored its effectiveness in various applications by Kim et al. 2023
+    (https://arxiv.org/abs/2310.02710).
+
+    Attributes:
+        estimator: the submitted PolicyEstimator for the forward pass.
+        pb_estimator: the PolicyEstimator for the backward pass.
+    """
+
+    def __init__(self, estimator: GFNModule, pb_estimator: GFNModule):
+        super().__init__(estimator)
+        self.backward_sampler = Sampler(pb_estimator)
+
+    def local_search(
+        self,
+        env: Env,
+        trajectories: Trajectories,
+        conditioning: torch.Tensor | None = None,
+        save_estimator_outputs: bool = False,
+        save_logprobs: bool = True,
+        back_steps: torch.Tensor | None = None,
+        back_ratio: float | None = None,
+        use_metropolis_hastings: bool = True,
+        **policy_kwargs: Any,
+    ) -> tuple[Trajectories, torch.Tensor]:
+        assert (
+            trajectories.log_rewards is not None
+        ), "Trajectories must have log rewards"
+        save_logprobs = save_logprobs or use_metropolis_hastings
+
+        device = trajectories.states.device
+        bs = trajectories.n_trajectories
+        state_shape = trajectories.states.state_shape
+        action_shape = trajectories.env.action_shape
+
+        # K-step backward sampling with the backward estimator,
+        # where K is the number of backward steps used in https://arxiv.org/abs/2202.01361.
+        if back_steps is None:
+            assert (
+                back_ratio is not None
+            ), "Either kwarg `back_steps` or `back_ratio` must be specified"
+            K = torch.ceil(back_ratio * (trajectories.when_is_done - 1)).long()
+        else:
+            K = torch.where(
+                back_steps > trajectories.when_is_done,
+                trajectories.when_is_done,
+                back_steps,
+            )
+
+        backward_trajectories = self.backward_sampler.sample_trajectories(
+            env,
+            states=trajectories.last_states,
+            conditioning=conditioning,
+            save_estimator_outputs=save_estimator_outputs,
+            save_logprobs=save_logprobs,
+            **policy_kwargs,
+        )
+
+        # Calculate the forward probability if needed (Metropolis-Hastings).
+        prev_trajectories = Trajectories.reverse_backward_trajectories(
+            backward_trajectories
+        )
+        prev_trajectories_log_rewards = trajectories.log_rewards
```

Comment:
should this be `prev_trajectories`? I actually think it does not matter since you're just looking at the reward at the end of the trajectory, but a comment here would be clarifying b/c above `prev_trajectories` refers to the reverse of a trajectory sampled from pf and here you're grabbing log_rewards directly from the forward trajectories.

Reply:
Why this is `prev_trajectories` -> check like 450-456. I wanted that part to be `new...` +/- `prev...`.
To alleviate the confusion, we can simply change this line to 
```
prev_trajectories_log_rewards = prev_trajectories.log_rewards
```